compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalgReduce.cpp

```diff
@@ -0,0 +1,692 @@
+#include "mlir/Transforms/DialectConversion.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace mlir::iree_compiler::stablehlo {
+namespace {
+namespace stablehlo = mlir::stablehlo;
+
+/// Returns a permutation AffineMap that puts all reduction dimensions to the
+/// last. The order of parallel loops and reduction loops are all sorted. E.g.,
+/// if `rank` is 4 and `reductionDims` is {1, 3}, then
+/// "(d0, d1, d2, d3) -> (d0, d2, d1, d3)" is used. The inverse permutation of
+/// the AffineMap is returned.
+AffineMap getTransposeMapForReduction(MLIRContext* context, int rank,
+                                      ArrayRef<int64_t> reductionDims) {
+  llvm::SmallSetVector<int, 4> s(reductionDims.begin(), reductionDims.end());
+
+  SmallVector<unsigned, 4> permutation;
+  for (int i = 0; i < rank; ++i)
+    if (!s.contains(i)) permutation.push_back(i);
+
+  llvm::append_range(permutation, reductionDims);
+  auto map = AffineMap::getPermutationMap(permutation, context);
+  return inversePermutation(map);
+}
+
+SmallVector<Value, 8> getReduceOpEmptyTensorDynSizes(
+    OpBuilder& b, Location loc, Value arg, ShapedType resultType,
+    ArrayRef<int64_t> reductionDims) {
+  llvm::SmallSetVector<int, 4> s(reductionDims.begin(), reductionDims.end());
+
+  SmallVector<unsigned, 4> parallelDims;
+  SmallVector<Value, 8> dynShape;
+  int rank = cast<RankedTensorType>(arg.getType()).getRank();
+  for (int i = 0, j = 0; i < rank; ++i) {
+    if (s.contains(i)) continue;
+    if (!resultType.isDynamicDim(j++)) continue;
+    dynShape.push_back(b.create<tensor::DimOp>(loc, arg, i));
+  }
+
+  return dynShape;
+}
+
+struct ReduceRegionReturnOpConversion final
+    : OpConversionPattern<stablehlo::ReturnOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      stablehlo::ReturnOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    if (!isInBodyOfLinalgOps(op)) {
+      return failure();
+    }
+
+    SmallVector<Value, 4> operands(adaptor.getOperands());
+    for (Value& operand : operands) {
+      if (isa<ShapedType>(operand.getType())) {
+        Location loc = operand.getLoc();
+        operand = rewriter.create<tensor::ExtractOp>(loc, operand);
+      }
+    }
+    rewriter.replaceOpWithNewOp<linalg::YieldOp>(op, operands);
+    return success();
+  }
+};
+
+struct ReduceOpToGenericConverter final
+    : OpConversionPattern<stablehlo::ReduceOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      stablehlo::ReduceOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    Location loc = op.getLoc();
+
+    int numOperands = static_cast<int>(adaptor.getInputs().size());
+
+    if (llvm::any_of(adaptor.getInputs(), [](Value v) {
+          return !v.getType().isa<RankedTensorType>();
+        })) {
+      return rewriter.notifyMatchFailure(op, "expects known-rank args");
```

Comment:
nit: weird formatting with braces here, could introduce a local variable to see if that helps clean it up

Reply:
This doesn't look unusual to me and passes clang-format, so I'll leave this as-is.